### PR TITLE
access log: consolidate and clean up address handling

### PR DESCRIPTION
**Pull request by mattklein123**
This is a wide reaching change that consolidates address handling
in request info / access logs to use real addresses vs. strings.
This will make it substantially easier to fill in protos when we
send gRPC access logs. It also has the added benefit of making
all address handling well typed. As part of this change I also
did the following:

- Added a new DOWNSTREAM_LOCAL_ADDRESS formatter.
- Made DOWNSTREAM_ADDRESS/CLIENT_IP print only the IP and not
  the port for IP addresses. This was done because of back-compat
  reasons for HTTP access logs. This will be a behavior change for
  TCP access logs. I don't see any way around this unfortunately
  without doing a ton of extra ugly work. Given that the % of users
  of HTTP logs dwarfs the % of users of TCP logs and that generally
  users do not care about the port of the remote, I think this is OK.
  If people disagree we can reconsider.
- Was able to cleanup various string mutations and parsing in the
  connection manager utility path as we are now operating on
  real addresses in many cases.

*Risk Level*: Medium: This is a wide ranging change in generally tricky code. We
have very good test coverage but thorough review is recommended.

*Testing*: Existing tests as well as new tests.

*Docs Changes*: https://github.com/envoyproxy/data-plane-api/pull/352

*Release Notes*: Updated both release notes and deprecated.

### CODE REVIEW

**Comment by ggreenway** [comment]
I haven't reviewed yet, but I need the ability to log the downstream remote port in addition to the address.  I'm fine with it being another formatter, eg "% DOWNSTREAM_ADDRESS%:%DOWNSTREAM_PORT%", but removing it entirely is a regression for me.

**Comment by mattklein123** [comment]
> I haven't reviewed yet, but I need the ability to log the downstream remote port in addition to the address. I'm fine with it being another formatter, eg "% DOWNSTREAM_ADDRESS%:%DOWNSTREAM_PORT%", but removing it entirely is a regression for me.

Alright no problem, I will add a new field for you.

File: source/common/request_info/utility.h [hunk 1]
```diff
@@ -34,5 +34,17 @@ class ResponseFlagUtils {
   const static std::string RATE_LIMITED;
 };
 
+/**
+ * Utility class for RequestInfo.
+ */
+class Utility {
```
Comment (ggreenway) [inline source/common/request_info/utility.h]: Should this be a namespace instead of a class?

File: test/integration/http_integration.cc [hunk 2]
```diff
@@ -1010,7 +1010,7 @@ void HttpIntegrationTest::testUpstreamProtocolError() {
   FakeRawConnectionPtr fake_upstream_connection = fake_upstreams_[0]->waitForRawConnection();
   // TODO(mattklein123): Waiting for exact amount of data is a hack. This needs to
   // be fixed.
-  fake_upstream_connection->waitForData(187);
+  fake_upstream_connection->waitForData(211);
```
Comment (ggreenway) [inline test/integration/http_integration.cc]: Can you explain what exactly the old and new value are here, and why it changed?  Or fix the TODO.

**Comment by mattklein123** [comment]
@ggreenway already fixed. See latest push. LMK if that works for you.

File: source/common/request_info/utility.h (hunk 1 above)
Comment (mattklein123) [reply source/common/request_info/utility.h]: I can change it. The rest of the code uses this pattern for the most part but I don't care either way if you have a strong preference.

File: test/integration/http_integration.cc (hunk 2 above)
Comment (mattklein123) [reply test/integration/http_integration.cc]: Sorry meant to drop a comment here. This change is due to the fact that when the connection manager does not use the remote address (attempts to pull from XFF), but XFF is absent or invalid, we use the real connection remote address. Thus, for loopback unit tests, we now set `x-envoy-internal: true` in certain cases which I think is more correct than what we had before.

File: source/common/request_info/utility.h (hunk 1 above)
Comment (ggreenway) [reply source/common/request_info/utility.h]: I don't feel strongly either way.

File: RAW_RELEASE_NOTES.md [hunk 3]
```diff
@@ -8,6 +8,11 @@ will make it substantially easier for the releaser to "linkify" all of the relea
 final version.
 
 ## 1.6.0
+* Added DOWNSTREAM_REMOTE_ADDRESS and DOWNSTREAM_REMOTE_ADDRESS_WITH_PORT access log formatters.
```
Comment (ggreenway) [inline RAW_RELEASE_NOTES.md]: I think this is backwards.  All the other address loggers (UPSTREAM_LOCAL_ADDRESS, DOWNSTREAM_LOCAL_ADDRESS) include ports.  I think DOWNSTREAM_REMOTE_ADDRESS should include the port, and then we can create DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT or something.  

Or alternately, we could try to make everything common and have *_ADDRESS_IP, *_ADDRESS_PORT, and then have users do "UPSTREAM_ADDRESS_IP:UPSTREAM_ADDRESS_PORT" in their format string; then they have have them together or separate.  But that raises more backwards-compat problems.  But whatever we come up with, it should be the most-consistent set of rules we can make.

Comment (mattklein123) [reply RAW_RELEASE_NOTES.md]: Yeah I had the same thought after I did this. Let me think about it and I will do another update.

**Comment by mattklein123** [comment]
@ggreenway I went with the "WITHOUT_PORT" option mainly because if we split it apart it doesn't handle the UDS case very well. I'm open to fully splitting into IP/PORT options, but not sure if that will be more confusing or less, given that UDS would return "-" or "" or whatever for port. I could go either way. LMK what you think.

**Comment by ggreenway** [comment]
It might make sense to split IP and port at some point, but it certainly doesn't need to happen in this PR.  I'm fine waiting until someone needs it/asks for it.

**Pull request merged**